### PR TITLE
IGNITE-13025: add IgniteLimit logical node

### PR DESCRIPTION
**Pull request by tledkov-gridgain**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-12407: Add Cluster API support to Java thin client`
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/LogicalRelImplementor.java [hunk 1]
```diff
@@ -17,7 +17,10 @@
 
 package org.apache.ignite.internal.processors.query.calcite.exec;
 
+import java.math.BigDecimal;
 import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
```
Comment (rkondakov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/LogicalRelImplementor.java:23]: Unused import.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/ExpressionFactoryImpl.java [hunk 2]
```diff
@@ -186,6 +193,29 @@ else if (collation.getFieldCollations().size() == 1)
         return SCALAR_CACHE.computeIfAbsent(digest(nodes, type), k -> compile(nodes, type));
     }
 
+    /** {@inheritDoc} */
+    @Override public <T> Supplier<CompletableFuture<T>> execute(RexNode node) {
```
Comment (rkondakov) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/ExpressionFactoryImpl.java:197]: Can't we return just a value here? Why do we need a) Supplier b) Future?

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/LimitOffsetTest.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+     *
+     * @param lim Limit.
+     * @param off Offset.
+     * @param param If {@code false} place limit/offset as literals. Otherwise they are plase as parameters.
+     * @param sorted Use sorted query (adds ORDER BY).
+     */
+    void checkQuery(int lim, int off, boolean param, boolean sorted) {
+        QueryEngine engine = Commons.lookupComponent(grid(0).context(), QueryEngine.class);
+
+        String sql = createSql(lim, off, false, false);
```
Comment (rkondakov) [inline modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/LimitOffsetTest.java:102]: `param, sorted ` instead of `false, false`?

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/LimitOffsetTest.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+        assertEquals("Invalid results size. [limit=" + lim + ", off=" + off + ", res=" + res.size() + ']',
+            expectedSize(lim, off), res.size());
+    }
+
+    /**
+     * Calculates expected result set size by limit and offset.
+     */
+    private int expectedSize(int lim, int off) {
+        if (off < 0)
```
Comment (rkondakov) [inline modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/LimitOffsetTest.java:119]: Should we throw an exception when limit or offset is negative? Or it's a valid value?

Comment (tledkov-gridgain) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/LimitOffsetTest.java:119]: At the test i use negative value not use limit/offset in the query (see #createSql). 
I'll add new test to check negative limit offset

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/LimitNode.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        if (!ended && rowNum >= offset) {
+            requested--;
+
+            downstream.push(row);
+        }
+
+        rowNum++;
+
+        if (requested > 0 && limit > 0 && rowNum >= limit + offset)
+            end();
```
Comment (gvvinblade) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/LimitNode.java:147]: let's cancel upstream only here and do not do it inside end() method

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/LimitNode.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    /** {@inheritDoc} */
+    @Override public void end() {
+        checkThread();
+
+        if (ended)
+            return;
+
+        ended = true;
+
+        sources.get(0).cancel();
```
Comment (gvvinblade) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/LimitNode.java:162]: We shouldn't cancel upstream here

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/Outbox.java [hunk 7]
```diff
@@ -157,24 +154,26 @@ public void init() {
         U.error(context().planningContext().logger(),
             "Error occurred during execution: " + X.getFullStackTrace(e));
 
-        cancel(); // TODO send cause to originator.
+        dest.targets().forEach(nodeId -> {
```
Comment (gvvinblade) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/Outbox.java:157]: let's move it into specific Buffer method

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/message/ErrorMessage.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+    public ErrorMessage() {
+        // No-op.
+    }
+
+    /** */
+    public ErrorMessage(UUID queryId, long fragmentId, long exchangeId, byte[] errBytes) {
+        this.queryId = queryId;
+        this.fragmentId = fragmentId;
+        this.exchangeId = exchangeId;
+        this.errBytes = errBytes;
```
Comment (gvvinblade) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/message/ErrorMessage.java:51]: Let's implement `MarshalableMessage` and put marshaling logic into appropriate methods

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/SortConverterRule.java [hunk 9]
```diff
@@ -40,12 +42,22 @@ public SortConverterRule() {
...
+        RelOptCluster cluster = sort.getCluster();
+        RelTraitSet outTraits = cluster.traitSetOf(IgniteConvention.INSTANCE).replace(sort.getCollation());
         RelTraitSet inTraits = cluster.traitSetOf(IgniteConvention.INSTANCE);
-        RelNode input = convert(rel.getInput(), inTraits);
+        RelNode input = convert(sort.getInput(), inTraits);
 
-        return new IgniteSort(cluster, outTraits, input, rel.getCollation(), rel.offset, rel.fetch);
+        IgniteRel res = null;
+
+        if (!sort.getCollation().getFieldCollations().isEmpty())
```
Comment (gvvinblade) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/SortConverterRule.java:53]: let's split into two rules: for sort and for limit (see `RelOptRule.matched()` method to separate rules better)

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeSystem.java [hunk 10]
```diff
@@ -20,10 +20,15 @@
...
 import org.apache.calcite.rel.type.RelDataTypeSystemImpl;
+import org.apache.calcite.sql.type.SqlTypeName;
 
 /**
  * Ignite type system.
  */
 public class IgniteTypeSystem extends RelDataTypeSystemImpl implements Serializable {
     public static final RelDataTypeSystem INSTANCE = new IgniteTypeSystem();
+
+    @Override public int getDefaultPrecision(SqlTypeName typeName) {
```
Comment (gvvinblade) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeSystem.java:31]: redundant overriding

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/LimitNode.java (hunk 5 above)
Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/LimitNode.java:147]: Good point. Fixed
